src/software/amazon/ionschema/internal/TypeImpl.kt

```diff
@@ -0,0 +1,70 @@
+package software.amazon.ionschema.internal
+
+import software.amazon.ion.*
+import software.amazon.ion.system.IonSystemBuilder
+import software.amazon.ionschema.Schema
+import software.amazon.ionschema.internal.constraint.ConstraintBase
+import software.amazon.ionschema.Violations
+
+/**
+ * Implementation of [Type] backed by a collection of zero or more [Constraint]s.
+ *
+ * If no "type" constraint is found, the default "type: any" is inserted by this class
+ * (unless addDefaultTypeConstraint is `false`).
+ */
+internal class TypeImpl(
+        private val ionStruct: IonStruct,
+        private val schema: Schema,
+        addDefaultTypeConstraint: Boolean = true
+    ) : TypeInternal, ConstraintBase(ionStruct) {
+
+    private companion object {
+        private val ION = IonSystemBuilder.standard().build()
+        private val ANY = ION.newSymbol("any")
+    }
+
+    internal val constraints: List<Constraint>
+
+    init {
+        var foundTypeConstraint = false
+        val tmpConstraints = ionStruct
```

Comment:
consider converting to a sequence before going through multiple processing steps, see https://blog.kotlin-academy.com/effective-kotlin-use-sequence-for-bigger-collections-with-more-than-one-processing-step-649a15bb4bf